server/log_rpc_server.go

```diff
@@ -504,31 +539,39 @@ func (t *TrillianLogRPCServer) GetEntryAndProof(ctx context.Context, req *trilli
 	if err != nil {
 		return nil, err
 	}
+	r := &trillian.GetEntryAndProofResponse{SignedLogRoot: &root}
 
-	proof, err := getInclusionProofForLeafIndex(ctx, tx, hasher, req.TreeSize, req.LeafIndex, root.TreeSize)
-	if err != nil {
-		return nil, err
+	if req.TreeSize > root.TreeSize && req.LeafIndex < root.TreeSize {
+		// return latest proof we can manage
```

Comment:
A client in this situation will either wait and try again, or request a proof against the tree head the server does know.

In the latter case, we save them at least one round trip. In the former, we cost them a little extra traffic.

Reply:
Yes but I think it shouldn't return the proof in this case, just the STH and let the client decide what to do.

Reply:
Well, I disagree. If it returns the STH, there's a risk that the next server is even more skewed and it will fail again. I realise that's pretty unlikely with current storage, but I can easily imagine other types of storage layers where it is far more likely.

Reply:
Not sure how this helps the client though. Maybe I'm not understanding what you mean? If client asks for proof at size 100 and gets one at size 50 for example. If 50 would have been enough then it could have been requested.

Reply:
That seems inefficient - client should request against an STH it already has, to reduce the number of STH consistency proofs it has to fetch...